File: Generated_test_cases/cpp/src/memory_pool.cpp
#include "memory_pool.h"
#include <cstdlib>
#include <cstring>

MemoryPool::MemoryPool(size_t poolSize) : pool_size(poolSize) {
    pool_start = static_cast<char*>(std::malloc(pool_size));
    used_memory = 0;
}

MemoryPool::~MemoryPool() {
    std::free(pool_start);
}

void* MemoryPool::alloc(size_t size) {
    if (free_list) {
        FreeNode* node = free_list;
        free_list = node->next;
        used_memory += size;
        return node;
    }
    if (used_memory + size <= pool_size) {
        void* ptr = pool_start + used_memory;
        used_memory += size;
        return ptr;
    }
    void* ptr = std::malloc(size);
    used_memory += size;
    return ptr;
}

void MemoryPool::free(void* ptr) {
    if (!ptr) return;
    FreeNode* node = static_cast<FreeNode*>(ptr);
    node->next = free_list;
    free_list = node;
}

void* MemoryPool::allocBuffer(size_t input_size, size_t multiplier, size_t extra) {
    size_t total = input_size * multiplier + extra;
    return alloc(total);
}

size_t MemoryPool::getUsedMemory() {
    return used_memory;
}

File: Generated_test_cases/JS/src/memorypool.js
class MemoryPool {
    constructor(size) {
        this.buffer = Buffer.alloc(size);
        this.offset = 0;
        this.size = size;
    }

    alloc(n) {
        if (n <= 0) return null;
        const currentOffset = this.offset;
        this.offset += n;                 // VULN-4: unsynchronized write – race condition
        if (currentOffset + n > this.size) {
            // overflow or out of pool; allocate new Buffer (fallback)
            return Buffer.alloc(n);
        }
        return this.buffer.slice(currentOffset, currentOffset + n);
    }

    free(buf) {
        // no-op for simplicity; in real pool would track freed memory
    }

    // VULN-1: Integer overflow in size calculation
    allocBuffer(inputSize, multiplier, extra) {
        const total = inputSize * multiplier + extra; // overflow possible beyond Number.MAX_SAFE_INTEGER
        return this.alloc(total);
    }

    usedMemory() {
        return this.offset;
    }
}

module.exports = MemoryPool;